src/shadertools/lib/shader-cache.js

```diff
@@ -51,21 +51,73 @@ export default class ShaderCache {
   /**
    * Returns a compiled `VertexShader` object corresponding to the supplied
    * GLSL source code string, if possible from cache.
-   *
    * @param {WebGLRenderingContext} gl - gl context
    * @param {String} source - Source code for shader
    * @return {FragmentShader} - a compiled fragment shader, possibly from chache
    */
   getFragmentShader(gl, source) {
-    assert(typeof source === 'string', ERR_SOURCE);
+    assert(typeof source === 'string');
+    assert(this._compareContexts(gl, this.gl));
 
     let shader = this.fragmentShaders[source];
-    assert(!shader || shader.gl === gl, ERR_CONTEXT);
-
     if (!shader) {
       shader = new FragmentShader(gl, source);
       this.fragmentShaders[source] = shader;
     }
     return shader;
   }
+
+  // Retrive Shaders from cache if exists, otherwise create new instance.
+  getProgram(gl, opts) {
+    assert(this._compareContexts(gl, this.gl));
+    assert(typeof opts.vs === 'string');
+    assert(typeof opts.fs === 'string');
+    assert(typeof opts.id === 'string');
+
+    const cacheKey = this._getProgramKey(opts);
+    let program = this.programs[cacheKey];
+    if (program) {
+      this._resetProgram(program);
+      return program;
```

Comment:
What happens if someone uses a 64bit and 32bit layer at the same time? Will this return separate programs, or will both get the 64/32 bit?

Reply:
Program cache key includes both shader strings, so it will return separate versions.